Wikipedia/Code/NavigationStateController.swift

```diff
@@ -21,6 +21,78 @@ final class NavigationStateController: NSObject {
+    @objc func restoreLastArticle(for navigationController: UINavigationController, in moc: NSManagedObjectContext, with theme: Theme, completion: @escaping () -> Void) {
+        guard navigationController.viewControllers.first is UITabBarController else {
+            assertionFailure("Expected root view controller to be UITabBarController")
+            completion()
+            return
+        }
+        guard let navigationState = moc.navigationState else {
+            completion()
+            return
+        }
+        
+        self.theme = theme
+        
+        var articleViewControllers: [NavigationState.ViewController] = []
+        for viewController in navigationState.viewControllers {
+            if let articleViewController = topmostArticleViewControllerFromViewController(viewController) {
+                articleViewControllers.append(articleViewController)
+            }
+        }
+        
+        guard let last = articleViewControllers.last,
+        let articleInfo = last.info else {
+            completion()
+            return
+        }
+        
+        let articleKind = last.kind
+        
+        guard let articleURL = articleURL(from: articleInfo) else {
+            completion()
+            return
+        }
+        
+        let viewControllerToPush: ArticleViewController?
+        if articleKind == .random {
+            viewControllerToPush = RandomArticleViewController(articleURL: articleURL, dataStore: dataStore, theme: theme)
+        } else {
+            viewControllerToPush = ArticleViewController(articleURL: articleURL, dataStore: dataStore, theme: theme)
+        }
+        
+        guard let viewControllerToPush else {
+            completion()
+            return
+        }
+        
+        viewControllerToPush.isRestoringState = true
+        pushOrPresent(viewControllerToPush, navigationController: navigationController, presentation: .push)
+        completion()
+    }
+    
+    private func topmostArticleViewControllerFromViewController(_ viewController: NavigationState.ViewController) -> NavigationState.ViewController? {
+        
+        var articleViewController: NavigationState.ViewController?
+        switch (viewController.kind, viewController.info) {
+        case (.article, _):
+            articleViewController = viewController
+        case (.random, _):
+            articleViewController = viewController
+        default:
+            break
+        }
+        
+        var childArticleViewController: NavigationState.ViewController?
+        for child in viewController.children {
+            childArticleViewController = topmostArticleViewControllerFromViewController(child)
+        }
+        
+        return childArticleViewController ?? articleViewController
+    }
+    
+    /// Attempts to restore user's entire heirarchy, included presented modal navigation stacks. Currently not in use in preference of restoreLastArticle to improve loading times.
```

Comment:
@tonisevener Super minor: typos heirarchy → hierarchy, included → including.

Reply:
Ah sorry about that, this is fixed (also in other places I found).